nixos/homepage-dashboard: support structured config

jnsgruk opened the pull request:
## Description of changes

This moves the `homepage-dashboard` configuration more toward the RFC42 approach. Until now, all configuration was initially generated and placed into `/var/lib/homepage-dashboard`, where users would be required to manually update the files, or place them there seperately in their nix config.

This allows for more "native" configuration.

There is some minor awkwardness here to avoid breaking people. This change shouldn't break anyone, but will display a deprecation warning about the upcoming change (which I've set at 24.05). The logic is that if you don't specify any of the new config options, the previous logic stays in play (homepage-dashboard is configured manually from /var/lib/homepage-dashboard), but if you adjust any of the sections, then it goes into "managed config" mode and starts writing the files to `/etc/homepage-dashboard` according to the native nix config.

For reference, this is what my server's dashboard config looks like: https://github.com/jnsgruk/nixos-config/blob/1f96734e905a2d2383e723f21f58059eacee3eda/host/common/services/homepage/thor.nix.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jnsgruk committed:
nixos/homepage-dashboard: support structured config

jnsgruk committed:
nixos/homepage-dashboard: add breaking change notice to release notes

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3524

vt52 commented:
I'm not qualified to give any sort of review, but just wanted to voice my support for this change. Only a couple weeks ago I was experimenting with this package and it was clear pretty quickly that it would benefit from a more managed config regime.

SuperSandro2000 commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,241 @@ in
+        '';
+        # Defaults: https://github.com/gethomepage/homepage/blob/main/src/skeleton/bookmarks.yaml
+        example = [
+          {
+            Developer = [
+              { Github = [{ abbr = "GH"; href = "https://github.com/"; }]; }
+            ];
+          }
+          {
+            Social = [
+              { Reddit = [{ abbr = "RE"; href = "https://reddit.com/"; }]; }
+            ];
+          }
+          {
+            Entertainment = [
+              { YouTube = [{ abbr = "YT"; href = "https://youtube.com/"; }]; }
+            ];
+          }
```
```suggestion
          {
            Developer = [
              { Github = [{ abbr = "GH"; href = "https://github.com/"; }]; }
            ];
          }
          {
            Entertainment = [
              { YouTube = [{ abbr = "YT"; href = "https://youtube.com/"; }]; }
            ];
          }
```

The example is a bit to big for search.nixos.org

jnsgruk replied:
Fixed!

SuperSandro2000 commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,241 @@ in
+        '';
+        # Defaults: https://github.com/gethomepage/homepage/blob/main/src/skeleton/services.yaml
+        example = [
+          {
+            "My First Group" = [
+              {
+                "My First Service" = {
+                  href = "http://localhost/";
+                  description = "Homepage is awesome";
+                };
+              }
+            ];
+          }
+          {
+            "My Second Group" = [
+              {
+                "My Second Service" = {
+                  href = "http://localhost/";
+                  description = "Homepage is the best";
+                };
+              }
+            ];
+          }
+          {
+            "My Third Group" = [
+              {
+                "My Third Service" = {
+                  href = "http://localhost/";
+                  description = "Homepage is 😎";
+                };
+              }
+            ];
+          }
```
```suggestion
          {
            "My First Group" = [ {
              "My First Service" = {
                href = "http://localhost/";
                description = "Homepage is awesome";
              };
            } ];
          }
          {
            "My Second Group" = [ {
              "My Second Service" = {
                href = "http://localhost/";
                description = "Homepage is the best";
              };
            } ];
          }
```

jnsgruk replied:
Fixed!

SuperSandro2000 commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,241 @@ in
+
+          options.logpath = lib.mkOption {
+            type = lib.types.str;
+            default = "/var/log/homepage-dashboard/homepage.log";
```
Can we just log to /dev/log instead? Do we otherwise need logrotate?

adamcstephens replied:
Or can the app just write to stdout as modern applications do?

jnsgruk replied:
I decided to drop this entirely, but also needed to add a patch to the upstream package. The app defaults to logging the same info to both a file and stdout as it's designed to be used in a Docker container with a volume mount. The latest commit just logs to stdout so that it's picked up in the journal, and the app doesn't try to create config directories in awkward places.

SuperSandro2000 commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,241 @@ in
+        + "bookmarks, services, widgets, and other configuration using the options provided.";
+    in
+    lib.mkIf cfg.enable {
+      warnings = if managedConfig then [ ] else [ msg ];
```
```suggestion
      warnings = lib.optional (!managedConfig) msg;
```

jnsgruk replied:
Fixed!

SuperSandro2000 commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,241 @@ in
+
+        environment = {
+          HOMEPAGE_CONFIG_DIR = configDir;
+          PORT = "${toString cfg.listenPort}";
```
```suggestion
          PORT = toString cfg.listenPort;
```

jnsgruk replied:
Fixed!

SuperSandro2000 commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,241 @@ in
+          EnvironmentFile = lib.mkIf (cfg.environmentFile != null) cfg.environmentFile;
+          StateDirectory = lib.mkIf (!managedConfig) "homepage-dashboard";
+          LogsDirectory = lib.mkIf managedConfig "homepage-dashboard";
+          ExecStart = "${lib.getExe cfg.package}";
```
```suggestion
          ExecStart = lib.getExe cfg.package;
```

jnsgruk replied:
Fixed!

jnsgruk committed:
homepage-dashboard: patch to enable control of log targets

This piece of software is generally used (and documented upstream)
in containers. It is common practice for such applications to log
both to stdout, and to a file which can be volume-mounted in the
container.

Because the intention here is to start with systemd, it makes less
sense to have both, so this added patch stops the application
from trying to create a log directory alongside the config directory.

jnsgruk committed:
nixosTests.homepage-dashboard: test managed and unmanaged configs

jnsgruk commented:
@SuperSandro2000 @adamcstephens thanks for the review! I've addressed your comments, and in doing so ended up patching the package.

I also added an additional test to cover the behaviour.

jnsgruk commented:
@SuperSandro2000 I should add, I'll work on a patch upstream that will make the log file optional, which would remove the need for the patch, but in the mean time I think the patch is simple enough and shouldn't cause any issues.

jnsgruk commented:
@SuperSandro2000 @adamcstephens okay I had some time this evening to try and get a patch landed upstream. I originally opened a PR, but the project requires a feature request with at leat 10 upvotes (I've added the request [here](https://github.com/gethomepage/homepage/discussions/3067)).

For now, I'm using `fetchpatch` to grab the patch from my branch. This simplifies the logging quite a bit.

jnsgruk commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,217 @@ in
+        environment = {
+          HOMEPAGE_CONFIG_DIR = configDir;
+          PORT = toString cfg.listenPort;
+          LOG_TARGETS = lib.mkIf managedConfig "stdout";
```
Control through this env var is introduced through the patch to the package.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1500

adamcstephens commented on nixos/modules/services/misc/homepage-dashboard.nix:
```diff
@@ -25,31 +27,217 @@ in
+
+      configDir = if managedConfig then "/etc/homepage-dashboard" else "/var/lib/homepage-dashboard";
+
+      msg = "using unmanaged configuration for homepage-dashboard will be deprecated in 24.05. "
```
```suggestion
      msg = "using unmanaged configuration for homepage-dashboard is deprecated and will be removed in 24.05. "
```

jnsgruk replied:
Done, thanks!

jnsgruk commented:
Thanks @adamcstephens!
